src/core/ddsi/src/ddsi_plist.c

```diff
@@ -678,6 +681,104 @@ static bool print_data_representation (char * __restrict *buf, size_t * __restri
   return prtf (buf, bufsize, ")");
 }
 
+#ifdef DDS_HAS_TYPE_DISCOVERY
+
+static dds_return_t deser_type_information (void * __restrict dst, struct flagset *flagset, uint64_t flag, const struct dd * __restrict dd, struct ddsi_domaingv const * const gv)
+{
+  (void) gv;
+  size_t dstoff = 0;
+  uint32_t srcoff = 0;
+  unsigned char *buf;
+  dds_return_t ret = 0;
+
+  if (dd->bswap)
+  {
+    buf = ddsrt_memdup (dd->buf, dd->bufsz);
+    if (!dds_stream_normalize_data ((char *) buf, &srcoff, (uint32_t) dd->bufsz, dd->bswap, CDR_ENC_VERSION_2, DDS_XTypes_TypeInformation_desc.m_ops))
+    {
+      ret = DDS_RETCODE_BAD_PARAMETER;
+      goto err_normalize;
+    }
+  }
+  else
+    buf = (unsigned char *) dd->buf;
+
+  dds_istream_t is = { .m_buffer = buf, .m_index = 0, .m_size = (uint32_t) dd->bufsz, .m_xcdr_version = CDR_ENC_VERSION_2 };
+  ddsi_typeinfo_t const ** x = deser_generic_dst (dst, &dstoff, alignof (ddsi_typeinfo_t *));
+  *x = ddsrt_calloc (1, DDS_XTypes_TypeInformation_desc.m_size);
+  dds_stream_read (&is, (void *) *x, DDS_XTypes_TypeInformation_desc.m_ops);
+  *flagset->present |= flag;
+err_normalize:
+  if (dd->bswap)
+    ddsrt_free (buf);
+  return ret;
+}
+
+static dds_return_t ser_type_information (struct nn_xmsg *xmsg, nn_parameterid_t pid, const void *src, size_t srcoff, enum ddsrt_byte_order_selector bo)
+{
+  ddsi_typeinfo_t const * const * x = deser_generic_src (src, &srcoff, alignof (ddsi_typeinfo_t *));
+
+  dds_ostream_t os = { .m_buffer = NULL, .m_index = 0, .m_size = 0, .m_xcdr_version = CDR_ENC_VERSION_2 };
+  if (bo == DDSRT_BOSEL_LE)
+    dds_stream_writeLE ((dds_ostreamLE_t *) &os, (const void *) *x, DDS_XTypes_TypeInformation_desc.m_ops);
+  else if (bo == DDSRT_BOSEL_BE)
+    dds_stream_writeBE ((dds_ostreamBE_t *) &os, (const void *) *x, DDS_XTypes_TypeInformation_desc.m_ops);
+  else
+    dds_stream_write (&os, (const void *) *x, DDS_XTypes_TypeInformation_desc.m_ops);
+  char * const p = nn_xmsg_addpar_bo (xmsg, pid, os.m_index, bo);
+  memcpy (p, os.m_buffer, os.m_index);
+  dds_ostream_fini (&os);
+  return 0;
+}
+
+static dds_return_t valid_type_information (const void *src, size_t srcoff)
+{
+  ddsi_typeinfo_t const * const * x = deser_generic_src (src, &srcoff, alignof (ddsi_typeinfo_t *));
+  /* FIXME: add more checks for type ids */
```

Comment:
Perhaps you can elaborate a bit, even if just in a GitHub comment?

Reply:
Sure, and because this validation is pretty straightforward I decided to implemented these checks:
https://github.com/dennis-adlink/cyclonedds/commit/eda0edc365775081ea8ebd312d7c7e410b1ae0f8

But when testing this, I noticed that the `valid` function is never called for this parameter (I think because it is defined as a `QP` and not a `PP`). I'm not sure what's the best way to implement this, can I add a call to `op.f.valid` in `init_one_parameter`? Or call the validate function from `deser_type_information`?

Reply:
After studying the code in `ddsi_plist.c` a bit more, I think this is the right approach for calling the `valid` function for parameters: 906b1041959aec899d935f0c977a0146828346e6 @eboasson could you please check if this is correct?